File: python/209_minimum-size-subarray-sum.py
# O(n) solution with Pointer
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        min_len = float('inf')
        left, cur_sum = 0, 0

        for i in range(len(nums)):
            cur_sum += nums[i]

            while cur_sum >= target:
                min_len = min(min_len, i - left + 1)

                cur_sum -= nums[left]
                left += 1

        return min_len if min_len != float('inf') else 0


#---------------------------------------------------------------------------------------#


# O(nlogn) Solution with Binary Search
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0

        ans = float('inf')
        sums = [0] * (n + 1)
        sums[0] = 0

        # cumulative sum
        for i in range(1, n + 1):
            sums[i] = sums[i - 1] + nums[i - 1]

        i, j = 0, 1
        while j < (n + 1):
            while sums[j] - sums[i] >= target:
                ans = min(ans, j - i)
                i += 1
            j += 1

        return ans if ans != float('inf') else 0

File: go/209_minimum-size-subarray-sum.go
// O(n) solution with Pointer
func minSubArrayLen(target int, nums []int) int {
    ans := len(nums) + 1
    left := 0
    sum := 0
    
    for i, n := range nums {
        sum += n
        for sum >= target {
            ans = min(ans, i-left+1)
            sum -= nums[left]
            left += 1
        }
    }
    if ans == len(nums) + 1 {
        return 0
    } else {
        return ans
    }
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

/*---------------------------------------------------------------------------------------*/

// O(nlogn) Solution with Binary Search
func minSubArrayLen(target int, nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    sums := make([]int, len(nums)+1)
    for i := 1; i < len(sums); i++ {
        sums[i] = sums[i - 1] + nums[i - 1]
    }
    
    ans := len(nums) + 1
    
    for i := 1; i < len(sums); i++ {
        if sums[i] >= target {
            left := binarySearch(0, i-1, sums[i]-target, sums)
            ans = min(ans, i-left)
        }
    }
    
    if ans == len(nums) + 1{
        return 0
    }
    return ans
}

func binarySearch(left, right, target int, sums []int) int {
    for left < right {
        mid := (left + right) / 2
        if sums[mid] < target {
            left = mid + 1
        } else {
            right = mid
        }
    }
    if sums[left] > target {
        return left - 1
    }
    return left
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

